Make the Qnap sensor more resilient if server is not reachable

mrosseel opened the pull request:
## Description:

My Qnap NAS server is booted on a schedule, so it's not always on when HA restarts.
**Current behaviour** is that if the NAS is not online, a persistent notification is posted and there's an error in the log.

**Goal**: silence this error if the NAS is not reachable, and let HA retry the connection till the NAS is on.

What I did: 
- added a configuration boolean 'ALLOW_UNREACHABLE'
- checked that if the NAS is not found and this boolean is True, the notification is not triggered.
- raise PlatformNotReady so that HA will retry
 
**possible issues**: 
- added ALLOW_UNREACHABLE to const.py, I realise this might not be wanted but there are many other components having this config option but declaring it in their own .py file. If this PR is accepted, some cleanup is possible in these files.
- if the NAS goes offline again, the old values are kept so we risk looking at stale values. This is also the case in the current implementation.
- I've held of on the documentation as this is my first PR, first want to see if the direction is ok etc.
- There was no test for this file, and I haven't added any.

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: qnap
    host: 192.168.0.10
    ssl: false
    username: user
    password: mypassword
    allow_unreachable: True
    monitored_conditions:
      - status
      - cpu_usage
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

mrosseel committed:
add CONF_ALLOW_UNREACHABLE option, retry connection if allowed

mrosseel committed:
Merge remote-tracking branch 'upstream/dev' into qnap

mrosseel committed:
fix linting errors

pvizeli commented on homeassistant/components/sensor/qnap.py:
```diff
@@ -99,6 +101,7 @@
     vol.Optional(CONF_NICS): cv.ensure_list,
     vol.Optional(CONF_DRIVES): cv.ensure_list,
     vol.Optional(CONF_VOLUMES): cv.ensure_list,
+    vol.Optional(CONF_ALLOW_UNREACHABLE, default=False): cv.boolean,
```
Why you make that as option?

mrosseel replied:
hi, the reasoning is that if this setting is False (the default), this sensor keeps its old behaviour. This is for people who have their NAS on 24/7 and want to see the notification. Setting it to True hides the notification and uses the HA PlatformNotReady mechanism.

BUT: maybe this option should dissapear and the PlatformNotReady should become the default, ditching the notification, that's another option and I'm open to opinions on this.

fabaff replied:
It should be only `PlatformNotReady` and the should be covered/handled by `available()`. Adding an option to configure the behavior of the notification looks wired .

mrosseel replied:
Hi, ok If I understand correctly you would just drop the config option and the notification, and use PlatformNotReady as the only behaviour. Sounds good, I'll remove the config. Anything else that needs to be changed? Not familiar with "available()".

pvizeli replied:
Available property is simple: return False if the the device is not reachable (maybe the state fetch going wrong on update) and True if they is back.

frenck commented:
Could not find a matching documentation PR. Added `docs-missing` label.

mrosseel committed:
Merge remote-tracking branch 'upstream/dev' into qnap

mrosseel committed:
Removing the config option, just using PlatformNotReady

mrosseel commented:
- removed the configuration option
- removed the notification (old behaviour when qnap nas was not available)

The docs_missing label can probably be removed because from a user's perspective, nothing described in the docs has changed.

frenck commented:
@mrosseel Correct! Since the parameter has been removed.

fabaff reviewed (approved):
Thanks :bird:
